### PR TITLE
Fix the saturated addition panic inside IPA

### LINKED ISSUE #1285: IPA run fails on 1M records
We're getting a runtime error consistently 

```bash
/target/release/build/ipa-core-9d838337f81c45b0/out/protocol_gate.rs:180474:17:
unexpected narrow for ProtocolGate(/ipa_prf/attribution/aggregate/aggregate_chunk02/aggregate00/saturating_add/add) => ThirtyTwoBitStep(bit00)
stack trace:
   0: std::backtrace::Backtrace::create
   1: ipa_core::error::set_global_panic_hook::{{closure}}
   2: std::panicking::rust_panic_with_hook
   3: std::panicking::begin_panic_handler::{{closure}}
   4: std::sys_common::backtrace::__rust_end_short_backtrace
   5: _rust_begin_unwind
   6: core::panicking::panic_fmt
   7: <ipa_core::protocol::step::ProtocolGate as ipa_step::StepNarrow<ipa_core::protocol::boolean::step::ThirtyTwoBitStep>>::narrow
   8: ipa_core::protocol::ipa_prf::boolean_ops::addition_sequential::addition_circuit::{{closure}}
   9: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  10: <futures_util::stream::stream::then::Then<St,Fut,F> as futures_core::stream::Stream>::poll_next
  11: ipa_core::protocol::ipa_prf::aggregation::aggregate_values::{{closure}}
  12: ipa_core::protocol::ipa_prf::aggregation::aggregate_contributions::{{closure}}::{{closure}}
  13: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  14: ipa_core::protocol::ipa_prf::prf_sharding::attribute_cap_aggregate::{{closure}}::{{closure}}
  15: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  16: ipa_core::protocol::ipa_prf::oprf_ipa::{{closure}}
  17: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  18: <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll
  19: <futures_util::future::future::flatten::Flatten<Fut,<Fut as core::future::future::Future>::Output> as core::future::future::Future>::poll
  20: tokio::runtime::park::CachedParkThread::block_on
  21: tokio::runtime::context::runtime::enter_runtime
  22: tokio::runtime::context::runtime_mt::exit_runtime
  23: tokio::runtime::scheduler::multi_thread::worker::block_in_place
  24: ipa_core::query::executor::do_query::{{closure}}
  25: tokio::runtime::task::core::Core<T,S>::poll
  26: tokio::runtime::task::harness::Harness<T,S>::poll
  27: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
  28: tokio::runtime::scheduler::multi_thread::worker::Context::run
  29: tokio::runtime::context::scoped::Scoped<T>::set
  30: tokio::runtime::context::runtime::enter_runtime
  31: tokio::runtime::scheduler::multi_thread::worker::run
  32: tokio::runtime::task::core::Core<T,S>::poll
  33: tokio::runtime::task::harness::Harness<T,S>::poll
  34: tokio::runtime::blocking::pool::Inner::run
  35: std::sys_common::backtrace::__rust_begin_short_backtrace
  36: core::ops::function::FnOnce::call_once{{vtable.shim}}
  37: std::sys::pal::unix::thread::Thread::new::thread_start
  38: __pthread_joiner_wake
```

### PR DESCRIPTION
**Pull request by akoshelev**
This fixes #1285. It was a silly error where the code used 32 bit step, but the compact gate declaration pointed to 16 bit. I spent the majority of the time actually reproducing this error in a unit test and making compact gate work with in-memory infra to allow people to write more of them.

So this PR essentially brings a test to reproduce the issue, runs more things with compact gate enabled and adds yet another approval step to the CI. It also fixes compile errors that we had for long time with compact gate and in memory infrastructure. Those errors stayed in the code because we never ran it with compact gate

### CODE REVIEW

**Comment by akoshelev** [comment]
draft run for 1M: https://draft-mpc.vercel.app/query/view/tubby-mix2024-09-19T0723

File: ipa-core/src/protocol/ipa_prf/mod.rs [hunk 1]
```diff
@@ -721,3 +721,109 @@ pub mod tests {
...
+                    user_id: 12345,
+                    is_trigger_report: true,
+                    breakdown_key: 0,
+                    trigger_value: 255,
+                },
+                TestRawDataRecord {
+                    timestamp: 20,
+                    user_id: 12345,
+                    is_trigger_report: true,
+                    breakdown_key: 0,
```
Comment (eriktaubeneck) [inline ipa-core/src/protocol/ipa_prf/mod.rs:780]:
```suggestion
                breakdown_key: 0,
```

Comment (eriktaubeneck) [reply ipa-core/src/protocol/ipa_prf/mod.rs:780]: now I can't tell if this is a rendering issue on my phone or an actual typo...

File: ipa-core/src/protocol/ipa_prf/mod.rs [hunk 2]
```diff
@@ -721,3 +721,109 @@ pub mod tests {
...
+                TestRawDataRecord {
+                    timestamp: 20,
+                    user_id: 68362,
+                    is_trigger_report: true,
+                    breakdown_key: 1,
+                    trigger_value: 255,
+                },
+            ];
+            let dp_params = DpMechanism::NoDp;
+            let padding_params = PaddingParameters::relaxed();
```
Comment (eriktaubeneck) [inline ipa-core/src/protocol/ipa_prf/mod.rs:806]: doesn't this happen by default when testing?

File: ipa-core/src/protocol/ipa_prf/mod.rs [hunk 3]
```diff
@@ -721,3 +721,109 @@ pub mod tests {
...
+            ipa_prf::{oprf_ipa, oprf_padding::PaddingParameters},
+            step::{ProtocolGate, ProtocolStep},
+        },
+        test_executor::run,
+        test_fixture::{ipa::TestRawDataRecord, Reconstruct, Runner, TestWorld, TestWorldConfig},
+    };
+
+    #[test]
+    fn saturated_agg() {
+        const EXPECTED: &[u128] = &[0, 255, 255, 0, 0, 0, 0, 0];
```
Comment (eriktaubeneck) [inline ipa-core/src/protocol/ipa_prf/mod.rs:746]: if we're trying to test going from 2 bytes to 4, shouldn't we test some much bigger numbers?

File: ipa-core/src/secret_sharing/into_shares.rs [hunk 4]
```diff
@@ -41,7 +41,7 @@ where
     }
 }
 
-#[cfg(all(test, unit_test))]
+#[cfg(test)]
```
Comment (eriktaubeneck) [inline ipa-core/src/secret_sharing/into_shares.rs:44]: for my own understanding, what is the difference between `test` and `unit_test`?

File: ipa-core/src/secret_sharing/vector/transpose.rs [hunk 5]
```diff
@@ -614,6 +614,8 @@ impl_transpose_shares_bool_to_ba!(BA32, 32, 256, test_transpose_shares_bool_to_b
 impl_transpose_shares_bool_to_ba_small!(BA8, 8, 32, test_transpose_shares_bool_to_ba_8x32);
 // added to support HV = BA32 to hold results when adding Binomial noise
 impl_transpose_shares_bool_to_ba_small!(BA32, 32, 32, test_transpose_shares_bool_to_ba_32x32);
+// // Usage: IPA test case for saturated additions
+// impl_transpose_shares_bool_to_ba_small!(BA3, 3, 32, test_transpose_shares_bool_to_ba_3x32);
```
Comment (eriktaubeneck) [inline ipa-core/src/secret_sharing/vector/transpose.rs:618]: intentionally commented out?

**Review by eriktaubeneck — APPROVED** [review]
nice find, and great to be able to build more tests with compact gate.

**Comment by akoshelev** [comment]
wrong link for Draft (still does not work for malicious security): https://draft-mpc.vercel.app/query/view/such-mitt2024-09-19T1629

File: ipa-core/src/protocol/ipa_prf/mod.rs (hunk 2 above)
Comment (akoshelev) [reply ipa-core/src/protocol/ipa_prf/mod.rs:806]: not sure what you mean by that. We have to supply this padding_params parameter for `oprf_ipa`, so we need to pick an explicit value (Rust does not support default values for function parameters)

File: ipa-core/src/secret_sharing/into_shares.rs (hunk 4 above)
Comment (akoshelev) [reply ipa-core/src/secret_sharing/into_shares.rs:44]: mainly `unit_test` includes `descriptive_gate`: https://github.com/private-attribution/ipa/blob/main/ipa-core/build.rs#L48 which prevents these trait implementations from being used in compact gate tests

File: ipa-core/src/secret_sharing/vector/transpose.rs (hunk 5 above)
Comment (akoshelev) [reply ipa-core/src/secret_sharing/vector/transpose.rs:618]: late night PR :( thanks for catching it!

File: ipa-core/src/protocol/ipa_prf/mod.rs (hunk 2 above)
Comment (eriktaubeneck) [reply ipa-core/src/protocol/ipa_prf/mod.rs:806]: got it. I was thinking this was gated by feature flags, but I suppose that just means this is unavailable without that flag.

**Pull request merged**